cpp-package/example/inference/imagenet_inference.cpp

```diff
@@ -0,0 +1,597 @@
+            << "INFO:" << "Batch size = " << input_shape[0] << " for inference" << std::endl
+            << "INFO:" << "Accuracy: " << val_acc.Get() << std::endl
+            << "INFO:" << "Throughput: " << (1000 * num_inference_batches * input_shape[0] / ms)
+            << " images per second" << std::endl;
+}
+
+Predictor::~Predictor() {
+  if (executor) {
+    delete executor;
+  }
+  if (!benchmark_) {
+    delete val_iter;
+  }
+  MXNotifyShutdown();
+}
+
+/*
+ * Convert the input string of number into the vector.
+ */
+template<typename T>
+std::vector<T> createVectorFromString(const std::string& input_string) {
+    std::vector<T> dst_vec;
+    char *p_next;
+    T elem;
+    bool bFloat = std::is_same<T, float>::value;
+    if (!bFloat) {
+      elem = strtol(input_string.c_str(), &p_next, 10);
+    } else {
+      elem = strtof(input_string.c_str(), &p_next);
+    }
+
+    dst_vec.push_back(elem);
+    while (*p_next) {
+        if (!bFloat) {
+          elem = strtol(p_next, &p_next, 10);
+        } else {
+          elem = strtof(p_next, &p_next);
+        }
+        dst_vec.push_back(elem);
+    }
+    return dst_vec;
+}
+
+void printUsage() {
+    std::cout << "Usage:" << std::endl;
+    std::cout << "imagenet_inference --symbol_file <model symbol file in json format>" << std::endl
+              << "--params_file <model params file> " << std::endl
+              << "--dataset <dataset used to run inference> " << std::endl
+              << "--data_nthreads <default: 60> " << std::endl
+              << "--input_shape <shape of input image e.g \"3 224 224\">] " << std::endl
+              << "--rgb_mean <mean value to be subtracted on RGB channel e.g \"0 0 0\">"
+              << std::endl
+              << "--rgb_std <standard deviation on R/G/B channel. e.g \"1 1 1\"> " << std::endl
+              << "--batch_size <number of images per batch> " << std::endl
+              << "--num_skipped_batches <skip the number of batches for inference> " << std::endl
+              << "--num_inference_batches <number of batches used for inference> " << std::endl
+              << "--data_layer_type <default: \"float32\" "
+              << "choices: [\"float32\",\"int8\",\"uint8\"]>" << std::endl
+              << "--gpu  <whether to run inference on GPU, default: false>" << std::endl
+              << "--benchmark <whether to use dummy data to run inference, default: false>"
+              << std::endl;
+}
+
+int main(int argc, char** argv) {
+  std::string model_file_json;
+  std::string model_file_params;
+  std::string dataset("");
+  std::string input_rgb_mean("0 0 0");
+  std::string input_rgb_std("1 1 1");
+  bool use_gpu = false;
+  bool benchmark = false;
+  int batch_size = 64;
+  int num_skipped_batches = 0;
+  int num_inference_batches = 100;
+  std::string data_layer_type("float32");
+  std::string input_shape("3 224 224");
+  int seed = 48564309;
+  int shuffle_chunk_seed = 3982304;
+  int data_nthreads = 60;
```

Comment:
If this is not the recommened `num_threads` by default, please change to a proper one.

Reply:
Actually, `num_threads` for data decoding will be determined by two parameters. One is `data_nthreads` and another is user-defined `OMP_NUM_THREADS`env variable through calling `omp_get_num_procs()`, please refer to [iter_image_recordio_2.cc#L146](https://github.com/apache/incubator-mxnet/blob/master/src/io/iter_image_recordio_2.cc#L146). So here I want to set a large number to make use of all cores. Also, `data_nthreads = 60` is to align with the current [Python implementation](https://github.com/apache/incubator-mxnet/blob/master/example/quantization/imagenet_inference.py#L157).